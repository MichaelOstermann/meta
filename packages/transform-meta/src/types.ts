import type { Node } from "oxc-parser"
import type { Meta } from "./Meta"

export interface TransformMetaResolver {
    (node: Node, meta: Meta): void
}

export interface TransformMetaOptions {
    hmr?: boolean
    resolve: TransformMetaResolver | TransformMetaResolver[]
    getName?: (name: string, node: Node, meta: Meta) => string
    getPath?: (path: string, meta: Meta) => string
}
