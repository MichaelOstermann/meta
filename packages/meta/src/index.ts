export * from "./fallbackMeta"
export * from "./getMeta"
export * from "./getMetaStack"
export * from "./hasMeta"
export * from "./isMeta"
export * from "./stringifyMeta"
export * from "./types"
export * from "./withMeta"
