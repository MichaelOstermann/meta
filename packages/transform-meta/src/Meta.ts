import type { Node, Program } from "oxc-parser"
import type { TransformMetaOptions } from "./types"
import Path from "node:path"
import MagicString from "magic-string"
import { parseSync } from "oxc-parser"
import { walk } from "oxc-walker"
import { getCallExpressionName } from "./helpers/getCallExpressionName"
import { getNodeName } from "./helpers/getNodeName"

export class Meta {
    ast: Program
    code: string
    filePath: string
    ms: MagicString
    usedIds = new Set<string>()
    #bodyOffset = 0
    #ids = new Map<string, string>()
    #options: TransformMetaOptions
    #parentNodes = new Map<Node, Node>()

    constructor(code: string, filePath: string, options: TransformMetaOptions) {
        this.code = code
        this.filePath = filePath
        this.#options = options
        this.ms = new MagicString(code, { filename: filePath })
        this.ast = parseSync(filePath, code).program
        this.#walkAst()
    }

    generateId(name: string): string {
        let i = 0
        let id = name
        while (this.usedIds.has(id)) {
            id = `${name}${++i}`
        }
        this.#ids.set(name, id)
        this.usedIds.add(id)
        return id
    }

    getCallExpressionName(node: Node): string {
        return getCallExpressionName(node)
    }

    getMetaLine(node: Node): number {
        let line = 1
        for (let i = 0; i < node.start; i++) {
            if (this.code.charCodeAt(i) === 10) line++
        }
        return line
    }

    getMetaName(node: Node): string {
        const names: string[] = []
        const nodes = [...this.getParentNodes(node), node]
        for (const n of nodes) getNodeName(n, names)
        const name = names.filter(Boolean).join(".")
        return this.#options.getName?.(name, node, this) ?? name
    }

    getMetaPath(): string {
        const path = Path.relative(process.cwd(), Path.resolve(this.filePath))
        return this.#options.getPath?.(path, this) ?? path
    }

    getParentNodes(node: Node): Node[] {
        const nodes: Node[] = []
        let pivot = this.#parentNodes.get(node)
        while (pivot) {
            nodes.unshift(pivot)
            pivot = this.#parentNodes.get(pivot)
        }
        return nodes
    }

    injectCode(code: string): void {
        this.ms.appendRight(this.#bodyOffset, `\n${code}`)
    }

    injectHmrSetup(): string {
        if (this.#ids.has("hmr")) return this.#ids.get("hmr")!
        const hmrId = this.generateId("hmr")
        this.injectCode(`const ${hmrId} = import.meta.hot ? (import.meta.hot.data["@monstermann/meta"] ??= new Set()) : undefined;`)
        this.injectCode(`if (${hmrId}) import.meta.hot.dispose(() => { for (const cb of ${hmrId}) { ${hmrId}.delete(cb); cb(); } });`)
        return hmrId
    }

    injectImport(code: string): void {
        this.ms.appendLeft(this.#bodyOffset, `\n${code}`)
    }

    injectMetaImport(name: string): string {
        if (this.#ids.has(name)) return this.#ids.get(name)!
        const id = this.generateId(name)
        if (id === name) this.injectImport(`import { ${name} } from "@monstermann/meta";`)
        else this.injectImport(`import { ${name} as ${id} } from "@monstermann/meta";`)
        return id
    }

    injectMetaParam(node: Node, metaId: string, argPos?: number): void {
        if (node.type !== "CallExpression") return
        let argLen = node.arguments.length

        const prefix = (val: string) => argLen > 0 ? `, ${val}` : val
        const append = (val: string) => this.ms.appendRight(node.end - 1, prefix(val))

        if (argPos == null) {
            append(metaId)
        }
        else if (argLen < argPos) {
            for (let i = argLen; i < argPos - 1; i++) {
                append("undefined")
                argLen++
            }
            append(metaId)
        }
    }

    injectMetaPath(path: string): string {
        if (!this.#ids.has("path")) {
            const pathId = this.generateId("path")
            this.injectCode(`const ${pathId} = "${path}";`)
        }
        return this.#ids.get("path")!
    }

    injectMetaRecord({ line, name, path }: {
        line: number
        name: string
        path: string
    }): string {
        const metaId = this.generateId("meta")
        const pathId = this.injectMetaPath(path)
        const props: string[] = [`path: ${pathId}`, `line: ${line}`, `name: "${name}"`]
        if (this.#options.hmr) props.push(`hmr: ${this.injectHmrSetup()}`)
        this.injectCode(`const ${metaId} = { ${props.join(", ")} };`)
        return metaId
    }

    injectMetaWrapper(node: Node, metaId: string): void {
        const withMetaId = this.injectMetaImport("withMeta")
        this.ms.appendLeft(node.start, `${withMetaId}(${metaId}, () => `)
        this.ms.appendRight(node.end, ")")
    }

    #walkAst() {
        walk(this.ast, {
            enter: (node, parent) => {
                if (node.type === "Identifier") this.usedIds.add(node.name)
                if (node.type === "ImportDeclaration") this.#bodyOffset = node.end
                if (node.type === "ImportDeclaration" && node.source.value === "@monstermann/meta") {
                    for (const spec of node.specifiers) {
                        if (spec.type !== "ImportSpecifier") continue
                        if (spec.imported.type === "Identifier" && spec.local.type === "Identifier") {
                            this.#ids.set(spec.imported.name, spec.local.name)
                        }
                    }
                }
                if (parent) this.#parentNodes.set(node, parent)
            },
        })
    }
}
