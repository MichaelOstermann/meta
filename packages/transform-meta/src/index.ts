export * from "./Meta"
export * from "./transformMeta"
export * from "./types"
export * from "./utils/addMetaParam"
export * from "./utils/setMetaParam"
export * from "./utils/wrapWithMeta"
