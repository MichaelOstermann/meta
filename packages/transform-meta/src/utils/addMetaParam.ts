import type { TransformMetaResolver } from "../types"

export function addMetaParam(identifiers: string[]): TransformMetaResolver {
    return function (node, meta) {
        const name = meta.getCallExpressionName(node)
        if (!identifiers.includes(name)) return
        const id = meta.injectMetaRecord({
            line: meta.getMetaLine(node),
            name: meta.getMetaName(node),
            path: meta.getMetaPath(),
        })
        meta.injectMetaParam(node, id)
    }
}
