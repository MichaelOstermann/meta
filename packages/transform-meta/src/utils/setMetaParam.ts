import type { TransformMetaResolver } from "../types"

export function setMetaParam(identifiers: Record<string, number>): TransformMetaResolver {
    return function (node, meta) {
        const name = meta.getCallExpressionName(node)
        const pos = identifiers[name]
        if (pos == null) return
        const id = meta.injectMetaRecord({
            line: meta.getMetaLine(node),
            name: meta.getMetaName(node),
            path: meta.getMetaPath(),
        })
        meta.injectMetaParam(node, id, pos)
    }
}
