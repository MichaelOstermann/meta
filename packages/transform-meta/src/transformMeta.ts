import type { SourceMap } from "magic-string"
import type { TransformMetaOptions } from "./types"
import { walk } from "oxc-walker"
import { Meta } from "./Meta"

export function transformMeta(
    code: string,
    filePath: string,
    options: TransformMetaOptions,
): {
    code: string
    map: SourceMap
} | undefined {
    const meta = new Meta(code, filePath, options)
    const resolvers = Array.isArray(options.resolve) ? options.resolve : [options.resolve]

    walk(meta.ast, {
        enter(node) {
            for (const resolve of resolvers) {
                resolve(node, meta)
            }
        },
    })

    if (!meta.ms.hasChanged()) return

    return {
        code: meta.ms.toString(),
        get map() {
            return meta.ms.generateMap({
                hires: "boundary",
                includeContent: true,
                source: filePath,
            })
        },
    }
}
