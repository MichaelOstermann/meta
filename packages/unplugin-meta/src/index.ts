import type { MetaPluginOptions } from "./types"
import { transformMeta } from "@monstermann/transform-meta"
import { createUnplugin } from "unplugin"

export { addMetaParam, setMetaParam, wrapWithMeta } from "@monstermann/transform-meta"

export default createUnplugin<MetaPluginOptions>(({
    enforce,
    exclude,
    getName,
    getPath,
    hmr,
    include,
    resolve,
}) => {
    let isViteDevServer = false

    return {
        enforce,
        name: "unplugin-meta",
        transform: {
            filter: {
                id: {
                    exclude,
                    include: include || [/\.[jt]sx?$/],
                },
            },
            handler(code, id) {
                return transformMeta(code, id, {
                    getName,
                    getPath,
                    hmr: hmr ?? isViteDevServer,
                    resolve,
                })
            },
        },
        vite: {
            configResolved(config) {
                isViteDevServer = config.command === "serve"
            },
        },
    }
})
