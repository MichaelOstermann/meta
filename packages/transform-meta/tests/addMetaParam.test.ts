import { describe, expect, it } from "vitest"
import { transformMeta } from "../src/transformMeta"
import { addMetaParam } from "../src/utils/addMetaParam"

describe("addMetaParam", () => {
    it("should append simple call expressions", () => {
        const transformed = transformMeta(`
        const example = foo()
        const example = foo(bar)
        `, "source.ts", {
            resolve: addMetaParam(["foo"]),
        })
        expect(transformed?.code).toMatchSnapshot()
    })

    it("should append new expressions", () => {
        const transformed = transformMeta(`
        const example = new Foo()
        const example = new Foo(bar)
        `, "source.ts", {
            resolve: addMetaParam(["Foo"]),
        })
        expect(transformed?.code).toMatchSnapshot()
    })

    it("should append call expressions with nested member expressions", () => {
        const transformed = transformMeta(`
        const example = foo.bar.baz()
        const example = foo.bar.baz(bar)
        `, "source.ts", {
            resolve: addMetaParam(["foo.bar.baz"]),
        })
        expect(transformed?.code).toMatchSnapshot()
    })
})
