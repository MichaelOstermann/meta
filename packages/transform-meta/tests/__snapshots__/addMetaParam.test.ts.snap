// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addMetaParam > should append call expressions with nested member expressions 1`] = `
"
const path = "source.ts";
const meta = { path: path, line: 2, name: "example" };
const meta1 = { path: path, line: 3, name: "example" };
        const example = foo.bar.baz(meta)
        const example = foo.bar.baz(bar, meta1)
        "
`;

exports[`addMetaParam > should append new expressions 1`] = `
"
const path = "source.ts";
const meta = { path: path, line: 2, name: "example" };
const meta1 = { path: path, line: 3, name: "example" };
        const example = new Foo()
        const example = new Foo(bar)
        "
`;

exports[`addMetaParam > should append simple call expressions 1`] = `
"
const path = "source.ts";
const meta = { path: path, line: 2, name: "example" };
const meta1 = { path: path, line: 3, name: "example" };
        const example = foo(meta)
        const example = foo(bar, meta1)
        "
`;
